### DISCUSSION TITLE
JS Client Service access

### DISCUSSION THREAD

**Discussion by martijnlammers** [microsoft/CCF#3143] [Q&A] [answered]
I've set up a simple single node open network with 1 active member and I'm now trying to setup a client. I've already successfully made a python client following the documentation and can make http requests over TLS. 

I will be working with AngularJS in the near future and was wondering if ccf also had a Javascript package to make signed/user/anonymous requests and if not, how I could achieve this in JS?

**Comment by eddyashton** [comment, marked as answer]
Hi @mAlyanak. No, we don't have any clients written in JS. We think it should be possible to write clients in any language, since they're speaking a standard protocol - HTTP over TLS - with common library support. We've successfully communicated with CCF nodes from a variety of existing tools (curl, k6, vegeta, artillery), and written clients in several languages (C++, Python, C#), so JS should be easy too.

The only tricky case is submitting _signed_ requests, which must be compatible with a specific version of the [draft RFC](https://microsoft.github.io/CCF/main/use_apps/issue_commands.html?highlight=http%20signatures#signing). There are libraries that do this in many languages (including JS), but checking they implement the same spec version may not be trivial.

**Reply by martijnlammers** [reply]
I've copied some random https example I found online. ( Sinning I know )

> statusCode:  500
> headers:  { 'content-length': '80', 'content-type': 'application/json' }
> {"error":{"code":"InternalError","message":"Exception thrown while executing."}}

I seem to be capable of accessing my VM, however it does give me the error message above. 
Do you have any idea what could possibly throw this error or where I could look to debug this issue?

> // jsclient.js

> const https = require('https');
> const fs = require('fs');
>   //key: fs.readFileSync('./certificates/Martijn_privk.pem'),
>   //cert: fs.readFileSync('./certificates/Martijn_cert.pem'),
>   //body: '{"key": "hello1", "value":"hello"}',
> 
> const data = '{"key": "hello1", "value":"hello"}';
> 
> var options = {
>   hostname: 'partical.uksouth.cloudapp.azure.com',
>   ca: fs.readFileSync('./certificates/network_cert.pem'),
>   port: 443,
>   path: '/app/read',
>   method: 'GET',
>   body: data,
>   headers: {
>     'content-type': 'application/json',
>   }
> };
> 
> var req = https.request(options, function(res) {
>   console.log("statusCode: ", res.statusCode);
>   console.log("headers: ", res.headers);
> 
>   res.on('data', function(d) {
>     process.stdout.write(d);
>   });
> });
> req.end();
> 
> req.on('error', function(e) {
>   console.error(e);
> });
>

**Reply by eddyashton** [reply]
That response means there was an error during execution of the JS app. Can you try calling `GET /node/commit`, to test the client connection independently of the app logic?

Then to debug what the error during JS execution is, you'll need to look in the node output (the `out` file in the node directory, if you launched it with `sandbox.sh`, else wherever you've redirected the output of `cchost` to). It should contain some additional information about the error, including potentially a JS callstack.

**Reply by martijnlammers** [reply]
/node/commit does seem to work.

![image](https://user-images.githubusercontent.com/60792074/139283530-4ee79051-8a49-4cd4-b460-dd47ee6bb7b4.png)

I however cannot find my stdout file created by the node (cchost). Can I explicitly configure where I create this file?

**Reply by eddyashton** [reply]
The output from `cchost` is written to `stdout`, and can be redirected to a file from bash. So if you run `cchost ...`, the output will be in that terminal window. If you run `cchost ... > /tmp/my_node.out`, it will be written to `/tmp/my_node.out`. The `sandbox.sh` script redirects to `workspace/<label>_<node_idx>/out`, where the workspace directory and label can be overridden by arguments.

**Reply by martijnlammers** [reply]
Oh alright, I first assumed there would be a seperate file generated from running cchost besides the normal output on the terminal. Afaik my node output currently isn't showing me anything else besides initial startup output and a notification the network has been opened.

**Reply by eddyashton** [reply]
That's odd. It should print an exception message or callstack when producing that 500 error message.

Can you show me the implementation of that endpoint (both the JS function and `app.json` entry for `GET /app/read`) and I'll see if I can spot anything that would crash. I assume it's an anonymous endpoint, and not trying to use the caller identity, since you're not setting a client cert or key in the `options`?

**Reply by martijnlammers** [reply]
Yes, that's correct. I turned off all authentication policies on my endpoints just for sake of testing.

> // read.js
> import { typedKv } from '../node_modules/@microsoft/ccf-app/kv.js';
> import { json, string } from '../node_modules/@microsoft/ccf-app/converters.js';
> import '../node_modules/@microsoft/ccf-app/historical.js';
> import '../node_modules/@microsoft/ccf-app/endpoints.js';
> 
> const kv = typedKv('kv', string, json());
> function read(request) {
>     const data = request.body.json();
>     if (typeof (data['key']) != "string") {
>         return {
>             body: '{"status":"invalid key"}',
>             statusCode: 400
>         };
>     }
>     if (!kv.has(data['key'])) {
>         return {
>             body: '{"status":"key not found for sure"}',
>             statusCode: 400
>         };
>     }
>     return {
>         body: kv.get(data['key']),
>         statusCode: 200
>     };
> }
> 
> export { read };
> 

Sorry for the formatting, looks very unappealing without the indents.
I'm fairly certain the logic is fine as the python client does work.
App.json:

> {
>     "endpoints": {
>         "/read": {
>             "get": {
>                 "js_module": "endpoints/read.js",
>                 "js_function": "read",
>                 "forwarding_required": "never",
>                 "authn_policies": [],
>                 "mode": "readonly",
>                 "openapi": {
>                 }
>             }
>         },
>         "/write": {
>             "post": {
>             "js_module": "endpoints/write.js",
>             "js_function": "write",
>             "forwarding_required": "never",
>             "authn_policies": [],
>             "mode": "readwrite",
>             "openapi": {
>             }
>         }
>         },
>         "/aggregate": {
>             "get": {
>             "js_module": "endpoints/aggregate.js",
>             "js_function": "aggregate",
>             "forwarding_required": "never",
>             "authn_policies": [],
>             "mode": "readonly",
>             "openapi": {
>             }
>         }
>         }
>     }
> }

**Reply by martijnlammers** [reply]
Aha, at last I did end up getting an error in my node output

> 2021-10-28T15:43:42.466252Z        0   [info ] ../src/enclave/main.cpp:173          | All threads are ready!
> 2021-10-28T15:46:35.389883Z -0.001 0   [info ] ../src/node/rpc/frontend.h:481       | Service state is OPEN, now accepting user transactions
> 2021-10-28T15:52:02.543694Z -0.001 0   [info ] ../src/js/wrap.cpp:713               | SyntaxError: unexpected end of input
> 2021-10-28T15:52:02.543727Z -0.001 0   [info ] ../src/js/wrap.cpp:735               |     at <body>:1
>     at json (native)
>     at read (/endpoints/read.js:8)

It took several minutes for me to realise it updated

**Reply by eddyashton** [reply]
The callstack says the error is on line 8, so during:
```
const data = request.body.json();
```

Presumably that means the body is either not present, or can't be parsed as JSON. A few ideas:
- Some of the framework code strips out request bodies for `GET` requests, since the HTTP spec says they shouldn't be used. I'm not sure if we do this for the JS endpoints, but we might. I guess if it's possible to `GET` that from Python with a body, we must not be stripping the body.
- I can't find documentation for the `body` field in the `options` object, and it's not listed under the fields [here](https://nodejs.org/api/http.html#httprequestoptions-callback). Are you sure that's how a request body should be passed? The sample suggests there is a separate `req.write(postData)` to fill the body. If it can be set as a `body` field in `options`, it's possible there is some additional encoding happening (ie - it has decided this is a string, so converted it to the bytes `"{...}"` in the actual HTTP request body).
- You can add additional logging with `console.log` in the endpoints, which will be written to stdout. So you can log when the endpoint is entered, and log what the current value of `request.body` is before trying to parse it as JSON.

**Reply by martijnlammers** [reply]
Hmm yeah that has a very high probability to be the case. I'll immediately try scaffolding the sourcecodes (Client + endpoint) with output to find out what's actually happening, definitely should've done this earlier. I'll report back after my findings.

**Reply by martijnlammers** [reply]
I actually honestly have no idea what changed - evidently just wrong syntax on my side - but I just used the example in the link you've given me, changed the module from http to https, added my custom options and *poof* it magically worked. You can even see in the node output that the body is identical to the one before, however it doesn't throw an error this time.

# Output using original method

> 2021-10-28T16:52:37.034344Z -0.002 0   [info ] ../src/js/wrap.cpp:713               | {"key":"hello1"}
> 2021-10-28T16:52:37.034377Z -0.002 0   [info ] ../src/js/wrap.cpp:713               | {"key":"hello1"}
> 2021-10-28T16:52:45.407024Z -0.001 0   [info ] ../src/js/wrap.cpp:713               | SyntaxError: unexpected end of input
> 2021-10-28T16:52:45.407057Z -0.001 0   [info ] ../src/js/wrap.cpp:735               |     at <body>:1
>     at json (native)
>     at read (/endpoints/read.js:8)
> 
 # Output using second method
> . . .
> 2021-10-28T17:02:26.753350Z -0.002 0   [info ] ../src/js/wrap.cpp:713               | {"key":"hello1"}
> 2021-10-28T17:02:26.753382Z -0.002 0   [info ] ../src/js/wrap.cpp:713               | {"key":"hello1"}
> 

Thanks for the help again Eddy.
